modules/externalipam/ip_cache.rb

```diff
@@ -0,0 +1,131 @@
+require 'yaml'
+require 'json'
+require 'monitor'
+require 'concurrent'
+require 'time'
+require 'externalipam/ipam_helper'
+require 'singleton'
+
+module Proxy::Ipam
+  # Class for managing temp in-memory cache to prevent same IP's being suggested in race conditions
+  class IpCache
+    include Singleton
+    include Proxy::Log
+    include Proxy::Ipam::IpamHelper
+
+    DEFAULT_CLEANUP_INTERVAL = 60
+
+    def initialize
+      @m = Monitor.new
+      @ip_cache = {'': {}}
+      start_cleanup_task
+    end
+
+    def provider_name(provider)
+      @provider = provider
+    end
+
+    def get_cidr(group_name, cidr)
+      @ip_cache.dig(group_name.to_sym, cidr.to_sym)
+    end
+
+    def get_ip(group_name, cidr, mac)
+      @ip_cache.dig(group_name.to_sym, cidr.to_sym, mac.to_sym, :ip)
+    end
+
+    def ip_exists(group_name, cidr, ip)
```

Comment:
Ruby convention suggests you could name this:
```suggestion
    def ip_exists?(group_name, cidr, ip)
```